Adopt dynamicDowncast<> in TextIterator

annevk opened the pull request:
#### 3cf5460f7c582c3397aa6ad0a098222957aa94b3
<pre>
Adopt dynamicDowncast&lt;&gt; in TextIterator
<a href="https://bugs.webkit.org/show_bug.cgi?id=268219">https://bugs.webkit.org/show_bug.cgi?id=268219</a>

Reviewed by Chris Dumez.

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::fullyClipsContents):
(WebCore::isRendererReplacedElement):
(WebCore::hasDisplayContents):
(WebCore::hasVisibleTextNode):
(WebCore::TextIterator::handleTextNode):
(WebCore::TextIterator::handleReplacedElement):
(WebCore::shouldEmitTabBeforeNode):
(WebCore::shouldEmitNewlinesBeforeAndAfterNode):
(WebCore::shouldEmitExtraNewlineForNode):
(WebCore::maxOffsetIncludingCollapsedSpaces):
(WebCore::TextIterator::shouldRepresentNodeOffsetZero):
(WebCore::SimplifiedBackwardsTextIterator::handleFirstLetter):

Canonical link: <a href="https://commits.webkit.org/273624@main">https://commits.webkit.org/273624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ce1192cee484e90c25256deccf5afaa7bf5de6e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38847 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32478 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12101 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31172 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36681 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12728 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32050 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11158 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40093 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32793 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32591 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37120 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11376 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35209 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13101 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8212 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -1019,8 +1003,11 @@ bool TextIterator::shouldRepresentNodeOffsetZero()
+        return false;
+
+    if (CheckedPtr renderBlockFlow = dynamicDowncast<RenderBlockFlow>(*currentNode->renderer());
+        renderBlockFlow && !renderBlockFlow->height() && !is<HTMLBodyElement>(currentNode))
```
The style checker does not like this, but this seems reasonable to me?

cdumez replied:
The style checker is right, since your condition is on more than one line, we prefer to use curly braces in this case. Please add curly braces.

annevk replied:
If I make it
```
    if (CheckedPtr renderBlockFlow = dynamicDowncast<RenderBlockFlow>(*currentNode->renderer());
        renderBlockFlow && !renderBlockFlow->height() && !is<HTMLBodyElement>(currentNode)) {
        return false;
    }
```
the style checker is still not happy:
```
ERROR: Source/WebCore/editing/TextIterator.cpp:1009:  Multi line control clauses should use braces.  [whitespace/braces] [4]
```
I didn't see anything in https://webkit.org/code-style-guidelines/#braces that would help me out.

cdumez commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -816,16 +812,15 @@ bool TextIterator::handleReplacedElement()
 static bool shouldEmitTabBeforeNode(Node& node)
 {
-    CheckedPtr renderer = node.renderer();
+    CheckedPtr cell = dynamicDowncast<RenderTableCell>(node.renderer());
```
This is not identical to the previous check, which was doing:
```
bool isTableCell(const Node& node)
{
    auto* renderer = node.renderer();
    if (!renderer)
        return node.hasTagName(tdTag) || node.hasTagName(thTag);
    return renderer->isRenderTableCell();
}
```

In particular, you're no longer doing this part of the check:
```
    if (!renderer)
        return node.hasTagName(tdTag) || node.hasTagName(thTag);
```

Which seems like it could change behavior.

annevk replied:
That is a good catch and I hadn't appreciated that nuance. However, if there's no renderer it would always return false so I think what I did is still correct. In particular, this is the status quo:
```
if (!renderer || !isTableCell(node))
    return false;
```

cdumez replied:
Ah, that's fair. I missed that. I think we're good then.

annevk committed:
Adopt dynamicDowncast<> in TextIterator
https://bugs.webkit.org/show_bug.cgi?id=268219

Reviewed by Chris Dumez.

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::fullyClipsContents):
(WebCore::isRendererReplacedElement):
(WebCore::hasDisplayContents):
(WebCore::hasVisibleTextNode):
(WebCore::TextIterator::handleTextNode):
(WebCore::TextIterator::handleReplacedElement):
(WebCore::shouldEmitTabBeforeNode):
(WebCore::shouldEmitNewlinesBeforeAndAfterNode):
(WebCore::shouldEmitExtraNewlineForNode):
(WebCore::maxOffsetIncludingCollapsedSpaces):
(WebCore::TextIterator::shouldRepresentNodeOffsetZero):
(WebCore::SimplifiedBackwardsTextIterator::handleFirstLetter):

Canonical link: https://commits.webkit.org/273624@main
